### PR TITLE
[#10176] Fix student tables not responding to changes

### PR DESCRIPTION
**Pull request by t-cheepeng**
Part of #10176

**Outline of Solution**
For deleting all students from the course details page, the `sections` was not resetted before fetching the updated data from backend. However, we could also simply reset the `sections` variable directly in the frontend and update statistics to 0 since it is guaranteed; saving one backend call. Would that be better?

For single deletes, the `IterableDiffer` in student list component only tracked the changes made to sections but did not track changes made to the list of students inside the section. Added tracking of list of students. Also updated the statistics that were being shown on the respective pages.

For the enroll students page, moved the second subscription to when the first one is completed before updating the table

**Update**
- Added in a `StatisticsCalculatorService` that holds logic for calculating course statistics from list of students.
- Refactored `student-list-component` to accept an already flattened list of students instead of having the component itself detect changes and flatten the list.
- Removed the change detection code
- Updated components which use the `student-list-component` to pass in a flat list of students
- Added an additional flattening of data in the `search-component`. Probably better to do flattening in component instead of the `search-service` itself.

**Instructor Course Details Page**
g

gh

**Instructor Student List Page**
ghj

**Instructor Enroll Students Page**
dfdf

### CODE REVIEW

File: src/web/app/pages-instructor/student-list/student-list.component.ts [hunk 1]
```diff
@@ -71,19 +79,46 @@ export class StudentListComponent implements OnInit, DoCheck {
               private tableComparatorService: TableComparatorService,
               private ngbModal: NgbModal,
               private differs: IterableDiffers) {
-    this._differ = this.differs.find(this.sections).create();
+    this.sectionDataDiffer = this.differs.find(this.sections).create();
```
Comment (xpdavid) [inline src/web/app/pages-instructor/student-list/student-list.component.ts:82]: Personally, I don't like to touch the low-level change detection in Angular as this is prone to errors. Even this is done before, it is still necessary to have one `StudentListSectionData` outside and one `FlatStudentListData` inside?

In other word, can we just construct `FlatStudentListData` (actually we should call it as `StudentListRowModel`) outside?

File: src/web/services/statistics-calculator.service.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  providedIn: 'root',
+})
+export class StatisticsCalculatorService {
+
+  constructor() { }
+
+  /**
+   * Calculates the statistics for a course from a list of students in the course
+   */
+  calculateCourseStatistics(studentList: StudentListRowModel[]): CourseStatistics {
```
Comment (xpdavid) [inline src/web/services/statistics-calculator.service.ts:26]: I think you can put the method into `StudentsService` or `CourseService`.

File: src/web/app/pages-instructor/student-list/student-list.component.ts [hunk 3]
```diff
@@ -39,21 +38,16 @@ interface FlatStudentListData {
   templateUrl: './student-list.component.html',
   styleUrls: ['./student-list.component.scss'],
 })
-export class StudentListComponent implements OnInit, DoCheck {
+export class StudentListComponent implements OnInit {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/student-list/student-list.component.ts:41]: Let's take this chance and put the component into `src/web/app/components`

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 4]
```diff
@@ -101,51 +97,42 @@ export class InstructorStudentListPageComponent implements OnInit {
...
         .subscribe((instructorPrivilege: InstructorPrivilege) => {
-          const sectionData: StudentListSectionData = {
-            sectionName,
-            students,
-            isAllowedToViewStudentInSection : instructorPrivilege.canViewStudentInSections,
-            isAllowedToModifyStudent : instructorPrivilege.canModifyStudent,
-          };
-
-          courseTab.studentListSectionDataList.push(sectionData);
+          students.forEach((student: StudentListRowModel) => {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts:128]: The privilege is loaded per section so you don't want to apply the privilege to all students (i.e., you need to apply the privilege to student in the `sectionName`).

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 5]
```diff
@@ -101,51 +97,42 @@ export class InstructorStudentListPageComponent implements OnInit {
...
+            student.isAllowedToViewStudentInSection = instructorPrivilege.canViewStudentInSections;
+            student.isAllowedToModifyStudent = instructorPrivilege.canModifyStudent;
+          });
+
+          courseTab.studentList.push(...students);
+          courseTab.stats =
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts:134]: Many redundant calculations are going here (though it will provide correct result eventually). I think the correct way to do it is to construct `StudentListRowModel` when all students are loaded. After that, we can begin to load privilege for each section.

File: src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts [hunk 6]
```diff
@@ -1,5 +1,22 @@
 import { Component, Input, OnInit } from '@angular/core';
-import { SearchStudentsTable } from '../instructor-search-page.component';
+import { StudentListSectionData } from '../../student-list/student-list-section-data';
+import { StudentListRowModel } from '../../student-list/student-list.component';
+
+/**
+ * Search result for a list of sections containing students of a course
+ */
+export interface SearchStudentsTable {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts:8]: Is this used somewhere?

File: src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+  courseId: string;
+  sections: StudentListSectionData[];
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts:10]: Can we delete the type `StudentListSectionData`?

File: src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts [hunk 8]
```diff
@@ -70,18 +64,47 @@ export class InstructorSearchPageComponent implements OnInit {
         .pipe(finalize(() => this.loadingBarService.hideLoadingBar()))
         .subscribe((resp: [InstructorSearchResult, InstructorSearchResult]) => {
           this.commentTables = resp[0].searchCommentsTables;
-          this.studentTables = resp[1].searchStudentsTables;
+          const searchStudentsTable: SearchStudentsTable[] = resp[1].searchStudentsTables;
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts:67]: It looks like we should also change the search service also. Let's leave it to another PR. Please raise an issue afterward or continue do it afterward :P.

File: src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.spec.ts [hunk 9]
```diff
@@ -44,44 +44,48 @@ describe('InstructorSearchPageComponent', () => {
...
-        sections: [
+        courseId: 'test-exa.demo',
+        students: [
           {
-            sectionName: 'Tutorial Group 1',
+            name: 'tester',
+            team: '1',
+            email: 'tester@tester.com',
+            status: JoinState.JOINED,
+            sectionName: '1',
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.spec.ts:56]: The section name should be `'Tutorial Group 1'` for all?

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 10]
```diff
@@ -157,19 +148,18 @@ export class InstructorCourseDetailsPageComponent implements OnInit {
...
     }).subscribe((instructorPrivilege: InstructorPrivilege) => {
-      const sectionData: StudentListSectionData = {
-        sectionName,
-        students,
-        isAllowedToViewStudentInSection : instructorPrivilege.canViewStudentInSections,
-        isAllowedToModifyStudent : instructorPrivilege.canModifyStudent,
-      };
-
-      this.sections.push(sectionData);
+      students.forEach((student: StudentListRowModel) => {
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts:156]: Same here. The privilege is not loaded correctly for each section.

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 11]
```diff
@@ -157,19 +148,18 @@ export class InstructorCourseDetailsPageComponent implements OnInit {
...
+        student.isAllowedToViewStudentInSection = instructorPrivilege.canViewStudentInSections;
+        student.isAllowedToModifyStudent = instructorPrivilege.canModifyStudent;
+      });
+
+      this.students.push(...students);
+      this.courseDetails.stats = this.statisticsCalculatorService.calculateCourseStatistics(this.students);
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts:162]: Same here. The stats calculation should be not here.

File: src/web/services/statistics-calculator.service.ts (hunk 2 above)
Comment (xpdavid) [reply src/web/services/statistics-calculator.service.ts:26]: In addition, let's not couple the model data and the business logic (services) together. Can we take `Student[]` as arguments?

File: src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts (hunk 8 above)
Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-search-page/instructor-search-page.component.ts:67]: Okay! Will raise and issue and work on it after this PR

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 12]
```diff
@@ -95,57 +89,53 @@ export class InstructorStudentListPageComponent implements OnInit {
...
-            numOfTeams: Object.keys(teams).length,
-          };
-
+          const studentsInCourse: Student[] = [];
           Object.keys(sections).forEach((key: string) => {
             const studentsInSection: Student[] = sections[key];
+            studentsInCourse.push(...studentsInSection);
 
-            const data: StudentListStudentData[] = [];
+            const studentList: StudentListRowModel[] = [];
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts:104]: There is a cleaner way of writing this by using `map`.

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 13]
```diff
@@ -327,11 +319,7 @@ export class InstructorCourseDetailsPageComponent implements OnInit {
     this.courseService.removeStudentFromCourse(this.courseDetails.course.courseId, studentEmail).subscribe(() => {
       this.statusMessageService
           .showSuccessMessage(`Student is successfully deleted from course "${this.courseDetails.course.courseId}"`);
-      this.sections.forEach(
-        (section: StudentListSectionData) => {
-          section.students = section.students.filter(
-            (student: StudentListStudentData) => student.email !== studentEmail);
-        });
+      this.loadStudents(this.courseDetails.course.courseId);
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts:322]: Same here

File: src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts [hunk 14]
```diff
@@ -195,7 +187,7 @@ export class InstructorCourseDetailsPageComponent implements OnInit {
   deleteAllStudentsFromCourse(courseId: string): void {
     this.studentService.deleteAllStudentsFromCourse({ courseId })
       .subscribe((resp: MessageOutput) => {
-        this.loadCourseDetails(courseId);
+        this.loadStudents(this.courseDetails.course.courseId);
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-course-details-page/instructor-course-details-page.component.ts:190]: If we delete all student, we can actually just clear the list (by setting to `[]`)

File: src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts [hunk 15]
```diff
@@ -1,5 +1,22 @@
...
-import { SearchStudentsTable } from '../instructor-search-page.component';
+import { StudentListSectionData } from '../../../components/student-list/student-list-section-data';
+import { StudentListRowModel } from '../../../components/student-list/student-list.component';
+
+/**
+ * Search result for a list of sections containing students of a course
+ */
+export interface SearchStudentsTable {
+  courseId: string;
+  sections: StudentListSectionData[];
```
Comment (xpdavid) [inline src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts:10]: Let's make sure that `StudentListSectionData` is only used in the search service.

Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts:10]: Yup. `StudentListSectionData` is only used in search service and its test files

Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-search-page/student-result-table/student-result-table.component.ts:10]: It is used once in `InstructorSearchPageComponent` because the search service currently returns `StudentListSectionData` and it is flattened in the component. Is that okay?

File: src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts [hunk 16]
```diff
@@ -158,11 +148,7 @@ export class InstructorStudentListPageComponent implements OnInit {
     this.courseService.removeStudentFromCourse(courseTab.course.courseId, studentEmail).subscribe(() => {
       this.statusMessageService
           .showSuccessMessage(`Student is successfully deleted from course "${courseTab.course.courseId}"`);
-      courseTab.studentListSectionDataList.forEach(
-          (section: StudentListSectionData) => {
-            section.students = section.students.filter(
-                (student: StudentListStudentData) => student.email !== studentEmail);
-          });
+      this.loadStudents(courseTab);
```
[the remark this answers is not in the document]
Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-student-list-page/instructor-student-list-page.component.ts:151]: Now we changed `calculateCourseStatistics` to accept `Student[]`. But the components only store the model. So we store an additional `Student[]` variable in these components? or would constructing the `Student[]` from the `StudentListRowModel[]` each time a delete event happens be a better choice? Personally, prefer the former

File: src/web/app/components/student-list/student-list.component.ts [hunk 17]
```diff
@@ -11,16 +15,11 @@ import { JoinState, MessageOutput } from '../../../types/api-output';
...
-  StudentListStudentData,
-} from './student-list-section-data';
 
 /**
- * Flattened data which contains details about a student and their section.
- * The data is flattened to allow sorting of the table.
+ * Model of row of student data containing details about a student and their section.
  */
-interface FlatStudentListData {
+export interface StudentListRowModel {
```
Comment (xpdavid) [inline src/web/app/components/student-list/student-list.component.ts:22]: For your question:

I think you can actually define a new field: `student: Student` here (and delete other student related field). Store the model in the corresponding page and when stats is calculated. Use the `student` field to construct a list of student.

**Comment by t-cheepeng** [comment]
> 
> 
> LGTM. I'll trust you to verify that all changes to snapshots are intended.

Yup! Have went through the updated snapshots to make sure the changes are intended
